torch/distributed/checkpoint/state_dict.py

```diff
@@ -215,6 +215,8 @@ def recurse(module: nn.Module, curr_fqn: str) -> Generator:
         for name, obj in chain(
             module.named_buffers(recurse=False), module.named_parameters(recurse=False)
         ):
+            if name in module._non_persistent_buffers_set:
```

Comment:
I might have missed some discussion. Could you remind me why we use `named_buffers()` rather than some logic that relies only on the keys in the state dict itself?

Reply:
It will trigger all_gather for FSDP. Since many users still use FSDP not FSDP2, we will have to ensure no performance penalty for this API.

Reply:
I see. The issue is that both full and sharded state dict all-gather?

Reply:
yes